File: hps-detector.js
/**
 * Harmonic Product Spectrum (HPS) Pitch Detection
 * Frequency-domain method that identifies the fundamental by multiplying
 * downsampled copies of the magnitude spectrum. Harmonics align at the
 * fundamental, creating a dominant peak even when the fundamental is weak.
 *
 * Accuracy Strategy (two-stage):
 *   Stage 1 — HPS on a 32768-point FFT identifies WHICH peak is the fundamental
 *             (harmonic rejection). This doesn't need to be super precise.
 *   Stage 2 — DTFT (Discrete-Time Fourier Transform) evaluation at arbitrary
 *             sub-bin frequencies around the identified peak gives near-perfect
 *             frequency resolution (sub-0.01 Hz).
 *
 * The DTFT evaluates X(f) = Σ x[n]·w[n]·e^(-j·2π·f·n/fs) at any frequency f,
 * not just integer bin centers. Each evaluation costs O(N) where N=bufferSize,
 * but we only need 6 evaluations (two rounds of 3), so the total cost is tiny.
 *
 * Depends on: pitch-common.js (TunerDefaults, PitchDetector)
 */
class HpsDetector extends PitchDetector {
    constructor(numHarmonics = 5) {
        super();
        this.numHarmonics = numHarmonics;
        this.debug = false;
        this.debugData = null;

        // FFT size: 32768 for fine HPS harmonic alignment
        // At 48kHz → ~1.46 Hz per bin (for HPS peak identification only)
        this.fftSize = 32768;

        // Preallocate reusable buffers
        this._fftReal = new Float64Array(this.fftSize);
        this._fftImag = new Float64Array(this.fftSize);
        this._magnitude = new Float64Array(this.fftSize / 2);
        this._hps = new Float64Array(this.fftSize / 2);

        // Buffer for windowed input (used by DTFT refinement)
        this._windowedBuf = null;

        // Precomputed window (lazily initialized to match input size)
        this._window = null;
        this._windowSize = 0;
    }

    /**
     * Blackman-Harris 4-term window.
     * Main lobe is wider than Hanning but sidelobes are -92 dB (vs -43 dB),
     * giving much cleaner spectral peaks for precise interpolation.
     */
    _ensureWindow(size) {
        if (this._windowSize === size) return;
        this._window = new Float64Array(size);
        this._windowedBuf = new Float64Array(size);
        const N1 = size - 1;
        const a0 = 0.35875, a1 = 0.48829, a2 = 0.14128, a3 = 0.01168;
        for (let i = 0; i < size; i++) {
            const phi = 2 * Math.PI * i / N1;
            this._window[i] = a0 - a1 * Math.cos(phi) + a2 * Math.cos(2 * phi) - a3 * Math.cos(3 * phi);
        }
        this._windowSize = size;
    }

    /**
     * In-place radix-2 Cooley–Tukey FFT.
     */
    _fft(re, im) {
        const n = re.length;

        // Bit-reversal permutation
        for (let i = 1, j = 0; i < n; i++) {
            let bit = n >> 1;
            while (j & bit) { j ^= bit; bit >>= 1; }
            j ^= bit;
            if (i < j) {
                let tmp = re[i]; re[i] = re[j]; re[j] = tmp;
                tmp = im[i]; im[i] = im[j]; im[j] = tmp;
            }
        }

        // Butterfly stages
        for (let len = 2; len <= n; len *= 2) {
            const half = len >> 1;
            const angle = -2 * Math.PI / len;
            const wRe = Math.cos(angle);
            const wIm = Math.sin(angle);
            for (let i = 0; i < n; i += len) {
                let curRe = 1, curIm = 0;
                for (let j = 0; j < half; j++) {
                    const idx1 = i + j;
                    const idx2 = idx1 + half;
                    const uRe = re[idx1], uIm = im[idx1];
                    const vRe = re[idx2] * curRe - im[idx2] * curIm;
                    const vIm = re[idx2] * curIm + im[idx2] * curRe;
                    re[idx1] = uRe + vRe;
                    im[idx1] = uIm + vIm;
                    re[idx2] = uRe - vRe;
                    im[idx2] = uIm - vIm;
                    const tmpRe = curRe * wRe - curIm * wIm;
                    curIm = curRe * wIm + curIm * wRe;
                    curRe = tmpRe;
                }
            }
        }
    }

    /**
     * Compute multi-harmonic DTFT score at a candidate fundamental frequency.
     *
     *   S(f) = Σ_{k=1}^{numH} |DTFT(k·f)|²
     *
     * By evaluating DTFT at f, 2f, 3f simultaneously, this score is:
     * - Sharper:  the combined peak is narrower than any single harmonic peak
     * - Robust:   even if the fundamental is weak (E2 string), stronger upper
     *             harmonics dominate the sum and stabilise the peak location
     * - Natural:  frequency errors are measured relative to pitch (musical
     *             intervals), not in absolute Hz — achieving what log-frequency
     *             bins would provide
     *
     * Uses 3 harmonics for refinement (not 5) to reduce sensitivity to
     * guitar string inharmonicity (higher partials deviate more).
     *
     * O(numH · N) per evaluation. For numH=3, N=4096: ~49K FLOPs.
     */
    _multiHarmonicScore(f, sampleRate, numH) {
        const wb = this._windowedBuf;
        const n = this._windowSize;
        let totalPower = 0;

        for (let h = 1; h <= numH; h++) {
            const omega = 2 * Math.PI * h * f / sampleRate;
            const cosW = Math.cos(omega);
            const sinW = Math.sin(omega);
            let re = 0, im = 0;
            let cosN = 1, sinN = 0;

            for (let i = 0; i < n; i++) {
                const s = wb[i];
                re += s * cosN;
                im -= s * sinN;
                const newCos = cosN * cosW - sinN * sinW;
                sinN = sinN * cosW + cosN * sinW;
                cosN = newCos;
            }
            totalPower += re * re + im * im;
        }
        return totalPower;
    }

    /**
     * Least-squares parabolic fit to symmetric evaluation points.
     *
     * Given 2k+1 points at offsets {-k, -(k-1), ..., 0, ..., k-1, k} × step
     * with log-domain scores, fits y = a·x² + b·x + c via normal equations.
     * For symmetric x-values, S_1 = S_3 = 0, giving the clean closed-form:
     *   b = T_1 / S_2         (linear coefficient)
     *   a = (n·T_2 - S_2·T_0) / (n·S_4 - S_2²)   (quadratic coefficient)
     *   peak at x = -b/(2a)
     *
     * Returns the fractional offset (in units of step) of the parabola's peak,
     * or 0 if the fit fails (non-concave or degenerate).
     */
    _leastSquaresPeak(offsets, logScores) {
        const n = offsets.length;
        let S_2 = 0, S_4 = 0, T_0 = 0, T_1 = 0, T_2 = 0;

        for (let i = 0; i < n; i++) {
            const x = offsets[i], y = logScores[i];
            const x2 = x * x;
            S_2 += x2;
            S_4 += x2 * x2;
            T_0 += y;
            T_1 += x * y;
            T_2 += x2 * y;
        }

        const det = n * S_4 - S_2 * S_2;
        if (det === 0 || S_2 === 0) return 0;

        const a = (n * T_2 - S_2 * T_0) / det;
        const b = T_1 / S_2;

        // Must be concave (a < 0) for a valid maximum
        if (a >= 0) return 0;

        const peak = -b / (2 * a);

        // Sanity: peak should be within the evaluation range
        const maxOff = offsets[offsets.length - 1];
        if (Math.abs(peak) > maxOff + 0.5) return 0;

        return peak;
    }

    /**
     * Two-round multi-harmonic DTFT refinement.
     *
     * Round 1 — Coarse:  7 points at ±1.5 bins (step = 0.5 bin)
     *                    least-squares parabola → ~0.05 bin accuracy
     * Round 2 — Fine:    5 points at ±0.25 bins around R1 result (step = 0.125 bin)
     *                    least-squares parabola → ~0.005 bin accuracy
     *
     * At 32768 FFT / 48kHz:
     *   0.005 bin ≈ 0.007 Hz ≈ 0.1 cent at E2, 0.04 cent at E4
     *
     * Total cost: (7 + 5) × 3 harmonics × 4096 samples ≈ 150K FLOPs (~0.1 ms)
     */
    _dtftRefine(approxFreq, sampleRate) {
        const binHz = sampleRate / this.fftSize;
        const numH = 3; // harmonics for refinement

        // ── Round 1: Coarse — 7 points over ±1.5 bins ──
        const step1 = binHz * 0.5;
        const offsets1 = [-3, -2, -1, 0, 1, 2, 3];
        const logScores1 = new Array(7);
        for (let i = 0; i < 7; i++) {
            const f = approxFreq + offsets1[i] * step1;
            if (f <= 0) { logScores1[i] = -100; continue; }
            const score = this._multiHarmonicScore(f, sampleRate, numH);
            logScores1[i] = score > 0 ? Math.log(score) : -100;
        }
        const peak1 = this._leastSquaresPeak(offsets1, logScores1);
        const freq1 = approxFreq + peak1 * step1;

        // ── Round 2: Fine — 5 points over ±0.25 bins around Round 1 ──
        const step2 = binHz * 0.125;
        const offsets2 = [-2, -1, 0, 1, 2];
        const logScores2 = new Array(5);
        for (let i = 0; i < 5; i++) {
            const f = freq1 + offsets2[i] * step2;
            if (f <= 0) { logScores2[i] = -100; continue; }
            const score = this._multiHarmonicScore(f, sampleRate, numH);
            logScores2[i] = score > 0 ? Math.log(score) : -100;
        }
        const peak2 = this._leastSquaresPeak(offsets2, logScores2);
        const freq2 = freq1 + peak2 * step2;

        return { frequency: freq2 };
    }

    getPitch(buffer, sampleRate) {
        const bufLen = buffer.length;

        // RMS noise gate
        let rms = 0;
        for (let i = 0; i < bufLen; i++) {
            rms += buffer[i] * buffer[i];
        }
        rms = Math.sqrt(rms / bufLen);
        if (rms < 0.01) {
            if (this.debug) this.debugData = null;
            return -1;
        }

        // Prepare window & windowed buffer (shared between FFT and DTFT)
        this._ensureWindow(bufLen);
        for (let i = 0; i < bufLen; i++) {
            this._windowedBuf[i] = buffer[i] * this._window[i];
        }

        const N = this.fftSize;
        const re = this._fftReal;
        const im = this._fftImag;

        // Zero-pad windowed buffer into FFT input
        for (let i = 0; i < bufLen; i++) {
            re[i] = this._windowedBuf[i];
        }
        for (let i = bufLen; i < N; i++) {
            re[i] = 0;
        }
        im.fill(0);

        // ── Stage 1: FFT-based HPS for harmonic identification ──

        this._fft(re, im);

        // Magnitude spectrum (positive frequencies only)
        const halfN = N >> 1;
        const mag = this._magnitude;
        for (let i = 0; i < halfN; i++) {
            mag[i] = Math.sqrt(re[i] * re[i] + im[i] * im[i]);
        }

        // Frequency range → bin range
        const minBin = Math.max(1, Math.floor(TunerDefaults.MIN_FREQUENCY * N / sampleRate));
        const maxBin = Math.min(
            Math.floor(halfN / this.numHarmonics),
            Math.ceil(TunerDefaults.MAX_FREQUENCY * N / sampleRate)
        );

        // Harmonic Product Spectrum (log domain)
        const hps = this._hps;
        for (let i = 0; i < maxBin; i++) {
            hps[i] = mag[i] > 0 ? Math.log(mag[i]) : -100;
        }
        for (let h = 2; h <= this.numHarmonics; h++) {
            for (let i = minBin; i < maxBin; i++) {
                const idx = i * h;
                if (idx < halfN) {
                    hps[i] += mag[idx] > 0 ? Math.log(mag[idx]) : -100;
                }
            }
        }

        // Find HPS peak
        let peakBin = minBin;
        let peakVal = hps[minBin];
        for (let i = minBin + 1; i < maxBin; i++) {
            if (hps[i] > peakVal) {
                peakVal = hps[i];
                peakBin = i;
            }
        }

        // Octave-error check (sub-harmonic preference)
        const halfPeakBin = Math.round(peakBin / 2);
        if (halfPeakBin >= minBin) {
            let bestSubBin = halfPeakBin;
            let bestSubVal = hps[halfPeakBin];
            const searchRadius = Math.max(2, Math.round(halfPeakBin * 0.05));
            const lo = Math.max(minBin, halfPeakBin - searchRadius);
            const hi = Math.min(maxBin - 1, halfPeakBin + searchRadius);
            for (let k = lo; k <= hi; k++) {
                if (hps[k] > bestSubVal) {
                    bestSubVal = hps[k];
                    bestSubBin = k;
                }
            }
            if (peakVal - bestSubVal < 1.5) {
                peakBin = bestSubBin;
                peakVal = bestSubVal;
            }
        }

        // SNR gate: peak must be well above the noise floor
        const hpsValues = [];
        for (let i = minBin; i < maxBin; i++) {
            hpsValues.push(hps[i]);
        }
        hpsValues.sort((a, b) => a - b);
        const median = hpsValues[Math.floor(hpsValues.length / 2)];
        const snr = peakVal - median;

        if (snr < 4) {
            if (this.debug) this.debugData = null;
            return -1;
        }

        // ── Stage 2: DTFT refinement for precise frequency estimation ──
        //
        // The HPS told us approximately WHERE the fundamental is (peakBin).
        // Now we find the *exact* spectral peak in the original magnitude
        // spectrum near that bin, then use DTFT evaluation at sub-bin
        // frequencies for near-perfect precision.

        // Find the true spectral peak in the original spectrum near the HPS bin
        let refBin = peakBin;
        let refVal = mag[peakBin];
        const refLo = Math.max(minBin, peakBin - 3);
        const refHi = Math.min(maxBin - 1, peakBin + 3);
        for (let k = refLo; k <= refHi; k++) {
            if (mag[k] > refVal) {
                refVal = mag[k];
                refBin = k;
            }
        }

        // Convert to frequency and refine via DTFT
        const binHz = sampleRate / N;
        const approxFreq = refBin * binHz;
        const refinement = this._dtftRefine(approxFreq, sampleRate);
        const frequency = refinement ? refinement.frequency : approxFreq;

        // Capture debug data
        if (this.debug) {
            const magDisplayMax = Math.min(halfN, maxBin * this.numHarmonics);

            // HPS-only frequency (for comparison in debug info)
            let hpsInterpolatedBin = peakBin;
            if (peakBin > minBin && peakBin < maxBin - 1) {
                const s0 = hps[peakBin - 1], s1 = hps[peakBin], s2 = hps[peakBin + 1];
                const d = s0 - 2 * s1 + s2;
                if (d < 0) {
                    const adj = (s0 - s2) / (2 * d);
                    if (Math.abs(adj) < 1) hpsInterpolatedBin = peakBin + adj;
                }
            }
            const hpsFreq = hpsInterpolatedBin * binHz;

            this.debugData = {
                magnitudeSpectrum: new Float64Array(mag.subarray(0, magDisplayMax)),
                hpsSpectrum: new Float64Array(hps.subarray(0, maxBin)),
                peakBin,
                interpolatedBin: frequency / binHz,
                peakVal,
                snr,
                rms,
                frequency,
                hpsOnlyFreq: hpsFreq,
                refBin,
                sampleRate,
                fftSize: N,
                minBin,
                maxBin,
                numHarmonics: this.numHarmonics
            };
        }

        return frequency;
    }
}

File: PianoTunerPython/hps_detector.py
import numpy as np
import math

class TunerDefaults:
    MIN_FREQUENCY = 20
    MAX_FREQUENCY = 10000

class HpsDetector:
    def __init__(self, num_harmonics=8, dtft_harmonics=2, snr_threshold=0.8, rms_threshold=0.002, window_size_exp=12):
        self.num_harmonics = num_harmonics
        self.dtft_harmonics = dtft_harmonics
        self.snr_threshold = snr_threshold
        self.rms_threshold = rms_threshold
        self.window_size_exp = window_size_exp
        
        self.fft_size = 32768
        self._window_size = 0
        self._window = None

    def _ensure_window(self, size):
        if self._window_size == size:
            return
        # Blackman-Harris 4-term
        N1 = size - 1
        a0, a1, a2, a3 = 0.35875, 0.48829, 0.14128, 0.01168
        idx = np.arange(size)
        phi = 2 * np.pi * idx / N1
        self._window = a0 - a1 * np.cos(phi) + a2 * np.cos(2 * phi) - a3 * np.cos(3 * phi)
        self._window_size = size

    def _multi_harmonic_score(self, f, sample_rate, num_h, weights, windowed_buf):
        n = len(windowed_buf)
        total_power = 0
        for h in range(1, num_h + 1):
            w = weights[h - 1] if weights is not None else 1
            if w <= 0:
                continue
            omega = 2 * math.pi * h * f / sample_rate
            
            # Vectorized DTFT evaluation
            idx = np.arange(n)
            exponent = -1j * omega * idx
            dtft_val = np.sum(windowed_buf * np.exp(exponent))
            power = w * (dtft_val.real**2 + dtft_val.imag**2)
            
            total_power += power
        return total_power

    def _assess_harmonic_weights(self, mag, half_n, ref_bin, num_h):
        weights = np.zeros(num_h)
        for h in range(1, num_h + 1):
            b = int(round(h * ref_bin))
            if b >= half_n - 5:
                weights[h - 1] = 0
                continue
            
            peak = mag[b]
            
            # noise floor from 3-5 bins away
            noise_sum = 0
            noise_count = 0
            for d in range(3, 6):
                if b - d >= 0:
                    noise_sum += mag[b - d]
                    noise_count += 1
                if b + d < half_n:
                    noise_sum += mag[b + d]
                    noise_count += 1
                    
            noise = noise_sum / noise_count if noise_count > 0 else 1
            local_snr = peak / noise if noise > 0 else 0
            weights[h - 1] = max(0, local_snr - 1.5)
            
        if weights[0] < 1:
            weights[0] = 1
        return weights

    def _least_squares_peak(self, offsets, log_scores):
        n = len(offsets)
        offsets = np.array(offsets)
        log_scores = np.array(log_scores)
        
        S_2 = np.sum(offsets**2)
        S_4 = np.sum(offsets**4)
        T_0 = np.sum(log_scores)
        T_1 = np.sum(offsets * log_scores)
        T_2 = np.sum((offsets**2) * log_scores)
        
        det = n * S_4 - S_2 * S_2
        if det == 0 or S_2 == 0:
            return 0
            
        a = (n * T_2 - S_2 * T_0) / det
        b = T_1 / S_2
        
        if a >= 0:
            return 0
            
        peak = -b / (2 * a)
        # Check bounds
        max_off = offsets[-1]
        if abs(peak) > max_off + 0.5:
            return 0
        return peak

    def _dtft_refine(self, approx_freq, sample_rate, weights, windowed_buf):
        bin_hz = sample_rate / self.fft_size
        num_h = len(weights)
        
        # Round 1
        step1 = bin_hz * 0.5
        offsets1 = [-3, -2, -1, 0, 1, 2, 3]
        log_scores1 = []
        for p in offsets1:
            f = approx_freq + p * step1
            if f <= 0:
                log_scores1.append(-100)
                continue
            score = self._multi_harmonic_score(f, sample_rate, num_h, weights, windowed_buf)
            log_scores1.append(math.log(score) if score > 0 else -100)
            
        peak1 = self._least_squares_peak(offsets1, log_scores1)
        freq1 = approx_freq + peak1 * step1
        
        # Round 2
        step2 = bin_hz * 0.125
        offsets2 = [-2, -1, 0, 1, 2]
        log_scores2 = []
        for p in offsets2:
            f = freq1 + p * step2
            if f <= 0:
                log_scores2.append(-100)
                continue
            score = self._multi_harmonic_score(f, sample_rate, num_h, weights, windowed_buf)
            log_scores2.append(math.log(score) if score > 0 else -100)
            
        peak2 = self._least_squares_peak(offsets2, log_scores2)
        freq2 = freq1 + peak2 * step2
        
        return freq2

    def get_pitch(self, buffer, sample_rate):
        required_size = 2 ** self.window_size_exp
        if len(buffer) < required_size:
            required_size = len(buffer)
            
        buffer = buffer[:required_size]
        buf_len = len(buffer)
        
        rms = np.sqrt(np.mean(buffer**2))
        if rms < self.rms_threshold:
            return -1
            
        target_rms = 0.1
        gain = target_rms / rms
        
        self._ensure_window(buf_len)
        windowed_buf = buffer * gain * self._window
        
        # FFT Stage
        re = np.zeros(self.fft_size)
        re[:buf_len] = windowed_buf
        
        fft_out = np.fft.rfft(re, n=self.fft_size)
        mag = np.abs(fft_out)
        half_n = self.fft_size // 2
        
        min_bin = max(1, math.floor(TunerDefaults.MIN_FREQUENCY * self.fft_size / sample_rate))
        internal_max_freq = 10000
        max_bin = min(half_n - 1, round(internal_max_freq * self.fft_size / sample_rate))
        
        hps = np.full(half_n, -100.0)
        
        for i in range(min_bin, max_bin):
            max_h = min(self.num_harmonics, math.floor((half_n - 1) / i))
            if max_h < 2:
                continue
            
            s = 0
            count = 0
            for h in range(1, max_h + 1):
                idx = i * h
                m = mag[idx]
                if m > 0:
                    s += math.log(m)
                    count += 1
                    
            if count >= 2:
                hps[i] = s / count
                
        # Find HPS peak
        peak_bin = min_bin
        peak_val = hps[min_bin]
        for i in range(min_bin + 1, max_bin):
            if hps[i] > peak_val:
                peak_val = hps[i]
                peak_bin = i
                
        # Octave-error check
        half_peak_bin = round(peak_bin / 2)
        if half_peak_bin >= min_bin:
            best_sub_bin = half_peak_bin
            best_sub_val = hps[half_peak_bin]
            search_radius = max(2, round(half_peak_bin * 0.05))
            lo = max(min_bin, half_peak_bin - search_radius)
            hi = min(max_bin - 1, half_peak_bin + search_radius)
            
            for k in range(lo, hi + 1):
                if hps[k] > best_sub_val:
                    best_sub_val = hps[k]
                    best_sub_bin = k
                    
            sub_mag = mag[best_sub_bin]
            peak_mag = mag[peak_bin]
            if (peak_val - best_sub_val) < 0.8 and sub_mag > peak_mag * 0.3:
                peak_bin = best_sub_bin
                peak_val = best_sub_val
                
        # SNR Gate
        valid_hps = hps[min_bin:max_bin]
        median = np.median(valid_hps)
        snr = peak_val - median
        
        if snr < self.snr_threshold:
            return -1
            
        # True spectral peak near HPS bin
        ref_bin = peak_bin
        ref_val = mag[peak_bin]
        ref_lo = max(min_bin, peak_bin - 3)
        ref_hi = min(max_bin - 1, peak_bin + 3)
        for k in range(ref_lo, ref_hi + 1):
            if mag[k] > ref_val:
                ref_val = mag[k]
                ref_bin = k
                
        # Assess quality
        bin_hz = sample_rate / self.fft_size
        approx_freq = ref_bin * bin_hz
        
        num_ref_h = self.dtft_harmonics
        if approx_freq < 150:
            max_useful = math.floor((sample_rate * 0.4) / approx_freq)
            num_ref_h = min(max_useful, max(4, self.dtft_harmonics))
            
        weights = self._assess_harmonic_weights(mag, half_n, ref_bin, num_ref_h)
        freq = self._dtft_refine(approx_freq, sample_rate, weights, windowed_buf)
        return freq
